microcosm/config/model.py

```diff
@@ -74,17 +76,62 @@ class Requirement:
     A value type for configuration defaults that represents *expected* config.
 
     """
-    def __init__(self, type=str, default_value=None, mock_value=None, required=True, *args, **kwargs):
+    def __init__(
+        self,
+        type=str,
+        default_value=None,
+        mock_value=None,
+        required=True,
+        default_factory=None,
+        *args,
+        **kwargs,
+    ):
         """
         :param type: a type callable
         :param mock_value: a default value to use durint testing (only)
 
         """
         self.type = type
         self.default_value = default_value
+        self.default_factory = default_factory
         self.mock_value = mock_value
         self.required = required
 
+        # Warn when the user doesn't provide exactly one of
+        #
+        # [`required=True`, `default_value=...`]
+        #
+        # In the future this will be an error, but for now we just warn in
+        # those situations which could occur as of the time this was introduced
+        warned = False
```

Comment:
I'm not the biggest fan of this branching logic, it makes it slightly harder to figure out the requirements. I also know it's temporary, so I'm not going to spend too much time nitpicking it.

That said, it would be nice to have these combined (the warning messages are arguably the same).

Maybe something like:
```
if any([
    all([
        required,
        default_value is not None or default_factory is not None,
    ]),
    all([
        not required,
        default_value is None or default_factory is None,
    ]),
]):
```
It's also a tad strange that supplying `default_factory` clobbers both of these messages, so if you *were* using it, you could run into the exception below.

Reply:
Your code definitely looks cleaner (and I will 💯steal conditional style at some point), but for this case the problem is that it changes the behavior.  I want to only warn in situations that could occur today.  For any code using the newly introduced `default_factory`, I want to fail hard, so that we don't end up with new code ignoring the warning.  Hence the guard on `default_factory is None`.  I did refactor to make it a bit cleaner, though.  Lmk what you think